### PR TITLE
MB-12251 - Document prime user actions

### PR DESCRIPTION
**Pull request by michaelvaldes-truss**
## Summary

This PR should do the following:
1. Create a new user role, "prime"
2. Link client certs to users and create a new "prime" user for each client cert
3. Get prime user info in the move history fetcher
4. Display prime user info in the ModifiedBy component

One thing that probably needs extra scrutiny is the generation of uuids in the migrations. I tried to avoid using uuid_generate_v4() when I could, but what else can we use when we're generating data in a loop?

[this article](https://dp3.atlassian.net/wiki/spaces/MT/pages/1702526996/Spike+-+Capturing+Prime+User+Information+for+Move+History) explains more about the approach used.

## Setup to Run Your Code

<details>
<summary>💻 You will need to use three separate terminals to test this locally.</summary>

##### Terminal 1

Run the migrations

```sh
make db_dev_migrate
```

##### Terminal 2

Start the UI locally.

```sh
make client_run
```

##### Terminal 3

Start the Go server locally.

```sh
make server_run
```

</details>

### Additional steps

1. Send a prime request! You should test via the prime simulator AND with either postman or with the prime cli. Note the locator of the move that is updated.
2. Access the office app as a TXO and navigate to the move that was updated in step 1.
3. Navigate to the 'Move history' tab and verify that the 'Modified by' information matches the acceptance criteria.

## Verification Steps for Author

These are to be checked by the author.

- [ ] Tested in the Experimental environment (for changes to containers, app startup, or connection to data stores)
- [ ] Request review from a member of a different team.
- [ ] Have the Jira acceptance criteria been met for this change?

## Verification Steps for Reviewers

These are to be checked by a reviewer.

### Frontend

- [ ] User facing changes have been reviewed by design.
- [ ] There are no aXe warnings for UI.
- [ ] This works in [Supported Browsers and their phone views](https://github.com/transcom/mymove/tree/master/docs/adr/0016-Browser-Support.md) (Chrome, Firefox, IE, Edge).
- [ ] There are no new console errors in the browser devtools
- [ ] There are no new console errors in the test output
- [ ] If this PR adds a new component to Storybook, it ensures the component is fully responsive, OR if it is intentionally not, a wrapping div using the `officeApp` class or custom `min-width` styling is used to hide any states the would not be visible to the user.

### Backend

- [ ] Code follows the guidelines for [Logging](https://transcom.github.io/mymove-docs/docs/dev/contributing/backend/Backend-Programming-Guide#logging)
- [ ] The requirements listed in [Querying the Database Safely](https://transcom.github.io/mymove-docs/docs/dev/contributing/backend/Backend-Programming-Guide/#querying-the-database-safely) have been satisfied.

#### Any new migrations/schema changes:

- [ ] Follows our guidelines for [Zero-Downtime Deploys](https://transcom.github.io/mymove-docs/docs/dev/contributing/database/Database-Migrations#zero-downtime-migrations)
- [ ] Have been communicated to #g-database
- [ ] Secure migrations have been tested following the instructions in our [docs](https://transcom.github.io/mymove-docs/docs/dev/contributing/database/Database-Migrations#secure-migrations)

## Screenshots
image

### CODE REVIEW

File: pkg/services/move_history/move_history_fetcher.go [hunk 1]
```diff
@@ -336,21 +336,25 @@ func (f moveHistoryFetcher) FetchMoveHistory(appCtx appcontext.AppContext, param
...
 				OR roles.role_type = 'transportation_invoicing_officer'
 				OR roles.role_type = 'ppm_office_users'
 				OR role_type = 'services_counselor'
 				OR role_type = 'contracting_officer'
 				OR role_type = 'qae_csr')
-		LEFT JOIN office_users ON office_users.user_id = session_userid
+		LEFT JOIN (
+			SELECT 'Prime' AS prime_user_first_name
+			) prime_users ON roles.role_type = 'prime'
+				OR roles.role_type = 'prime_simulator'
```
Comment (rogeruiz) [inline pkg/services/move_history/move_history_fetcher.go:357]:
```suggestion
			) prime_users ON roles.role_type LIKE 'prime%'
```

What do you think about using this? That way we can match on any roles that start like `prime` and that would allow us to match on Prime roles we add in the future with the same name of `prime.+`.

File: migrations/app/schema/20220531203716_add_user_for_each_client_cert.up.sql [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+BEGIN
+	prime_role_id := (SELECT id FROM roles WHERE role_type = 'prime');
+
+	FOR client_cert IN
+	SELECT *
+	FROM client_certs
+		LOOP
+			-- Create a user for each client cert
+			new_user_id := uuid_generate_v4();
```
Comment (michaelvaldes-truss) [inline migrations/app/schema/20220531203716_add_user_for_each_client_cert.up.sql:15]: Is it okay that we're using uuid_generate_v4() here and below? If not, how else can deal with a scenario in which we need to add uuids in a loop?

File: pkg/services/move_history/move_history_fetcher.go [hunk 3]
```diff
@@ -336,21 +336,25 @@ func (f moveHistoryFetcher) FetchMoveHistory(appCtx appcontext.AppContext, param
...
-		office_users.first_name AS session_user_first_name,
+		COALESCE(office_users.first_name, prime_user_first_name) AS session_user_first_name,
 		office_users.last_name AS session_user_last_name,
 		office_users.email AS session_user_email,
 		office_users.telephone AS session_user_telephone
 	FROM
 		combined_logs
 		LEFT JOIN users_roles ON session_userid = users_roles.user_id
 		LEFT JOIN roles ON users_roles.role_id = roles.id
+		LEFT JOIN office_users ON office_users.user_id = session_userid
```
Comment (michaelvaldes-truss) [inline pkg/services/move_history/move_history_fetcher.go:347]: Is it okay to join by roles like this? Will it be okay when a user has multiple roles?

File: pkg/services/move_history/move_history_fetcher.go (hunk 1 above)
Comment (michaelvaldes-truss) [reply pkg/services/move_history/move_history_fetcher.go:357]: Sure, that works for me

File: pkg/services/move_history/move_history_fetcher.go (hunk 3 above)
Comment (donaldthai) [reply pkg/services/move_history/move_history_fetcher.go:347]: thought: I think those are fair assumptions. I definitely don't think there will be an overlap between office users and prime users.

File: migrations/app/schema/20220531203716_add_user_for_each_client_cert.up.sql (hunk 2 above)
Comment (michaelvaldes-truss) [reply migrations/app/schema/20220531203716_add_user_for_each_client_cert.up.sql:15]: heres the documentation for creating migrations
https://transcom.github.io/mymove-docs/docs/backend/setup/database-migrations#why-we-avoid-uuid_generate_v4

File: pkg/testdatagen/make_client_cert.go [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+	mergeModels(&clientCert, assertions.ClientCert)
+
+	existingCert, err := models.FetchClientCert(db, clientCert.Sha256Digest)
+	if err != nil && errors.Cause(err).Error() == models.ErrFetchNotFound.Error() {
+		// dev client cert not found; create new client cert
+		mustCreate(db, &clientCert, false)
+	} else if err == nil && existingCert != nil {
+		// client cert already exists; update existing client cert
+		mergeModels(&existingCert, clientCert)
+		MustSave(db, &existingCert)
```
Comment (donaldthai) [inline pkg/testdatagen/make_client_cert.go:41]: suggestion: I was running into an error when running the `db_dev_e2e_populate` make call. Noticed there the `existingCert` that is fetched is already in a pointer type so the `&` is not necessary here.

```suggestion
		mergeModels(existingCert, clientCert)
		MustSave(db, existingCert)
```

Comment (michaelvaldes-truss) [reply pkg/testdatagen/make_client_cert.go:41]: Thanks for catching that!

**Review by rogeruiz — APPROVED** [review]
lgtm 🌈

Paired with @michaelvaldes-truss on this. Everything is working fine and we were able to see the Prime in the Modified column. Also `make server_test` passed.

File: .circleci/config.yml [hunk 5]
```diff
@@ -30,10 +30,10 @@ references:
   #
   # In addition, it's common practice to disable acceptance tests and
   # ignore tests for dp3 deploys. See the branch settings below.
-  dp3-branch: &dp3-branch placeholder_branch_name
+  dp3-branch: &dp3-branch MB-12251-Document_prime_user_actions
   # MUST BE ONE OF: loadtest, demo, exp.
   # These are used to pull in env vars so the spelling matters!
-  dp3-env: &dp3-env placeholder_env
```
Comment (donaldthai) [inline .circleci/config.yml]: suggestion: Make sure to revert these changes back before merging

**Review by donaldthai — APPROVED** [review]
I think this looks good! 🤩 

I think the only thing I can think of that we might need is to alter the `user_id` to be not null after generating new users for each client certs.
